### ISSUE TITLE
[Draft] Scoped Future

### ISSUE THREAD

**Issue by ethe** [monoio-rs/monoio#126]
## Motivation
In current, [monoio::spawn](https://github.com/bytedance/monoio/blob/45d74ec2d40b50f843573c6744646a575b57e38e/monoio/src/runtime.rs#L362) requires `Future` and its output must be `'static`. However, user maybe takes lots of needless overhead (such as `Rc` to those captured objects) to promise `'static`. If we consider the `Future: ?Send + ?Sync` requirment of `monoio::spawn`, removing the `'static` requirement should be possible, because the lifetime of `JoinHandle` is able to be explicit and it can not be shared between threads. It is something like the future version of [scoped-threads](https://rust-lang.github.io/rfcs/3151-scoped-threads.html).

## Advantages
Scoped-future allows user use reference (`&'future _` / `&'future mut _`) rather than `Rc` or `Rc<RefCell<_>>` to those object which are captured in async block, it is useful to reduce the runtime overhead in some perfomance-sensitive cases.

## Possible Design
Add `scoped_spawn` function and it returns `ScopedJoinHandle`:
```rust
pub fn scoped_spawn<'future, T>(future: T) -> ScopedJoinHandle<'future, T::Output>
where
    T: Future + 'future,
    T::Output: 'future,
```
`ScopedJoinHandle` must be awaited or setted canceled in lifetime. In first case, maybe we are able to use `#[must_use]` lint to suggest users, and force deallocate the raw task in `<ScopedJoinHandle as Drop>::drop` in second case to prevent data race.

**Comment by MartyMcFlyInTheSky** [comment]
Yes that would be nice. Or just do something similar as thread::scoped or [tokio_scoped](https://docs.rs/tokio-scoped/latest/tokio_scoped/), that would already suffice.
